api/query_engine.go

```diff
@@ -15,84 +17,111 @@ var (
-	// the first (raw) uses the LCM of different raw intervals, since that's what needed to fulfill a raw request.
-	// in edge cases (poorly configured setups) this might make raw interval > 1st rollup
-	if archive == 0 {
-		seenRawIntervals := make(map[uint32]struct{})
-		for _, req := range reqs {
-			if _, ok := seenRawIntervals[req.RawInterval]; !ok {
-				listRawIntervals = append(listRawIntervals, req.RawInterval)
-				seenRawIntervals[req.RawInterval] = struct{}{}
-			}
+		if _, ok := seenIntervals[req.ArchInterval]; !ok {
+			listIntervals = append(listIntervals, req.ArchInterval)
+			seenIntervals[req.ArchInterval] = struct{}{}
 		}
-		interval = util.Lcm(listRawIntervals)
-	} else {
-		interval = s.Aggs[archive-1].Span
 	}
 
-	/* we now just need to update the following properties for each req:
-	   archive      int    // 0 means original data, 1 means first agg level, 2 means 2nd, etc.
-	   archInterval uint32 // the interval corresponding to the archive we'll fetch
-	   outInterval  uint32 // the interval of the output data, after any runtime consolidation
-	   aggNum       uint32 // how many points to consolidate together at runtime, after fetching from the archive
-	*/
+	// due to different retentions coming into play, different requests may end up with different resolutions
+	// we all need to emit them at the same interval, the LCM interval >= interval of the req
+
+	interval := util.Lcm(listIntervals)
+
+	// now, for all our requests, set all their properties.  we may have to apply runtime consolidation to get the
+	// correct output interval if out interval != native.  In that case, we also check whether we can fulfill
+	// the request by reading from an archive instead (i.e. whether it has the correct interval.
+	// the TTL of lower resolution archives is always assumed to be at least as long so we don't have to check that)
 
-	// only apply runtime consolidation (pre data processing in graphite api) if we have to due to non uniform raw intervals
-	runtimeConsolidate := archive == 0 && len(listRawIntervals) > 1
 	var pointsFetch uint32
 	for i := range reqs {
 		req := &reqs[i]
-		req.Archive = archive
-		req.TTL = ttl
-		req.OutInterval = interval
-
-		req.ArchInterval = interval
-		req.AggNum = 1
-		if runtimeConsolidate {
-			req.ArchInterval = req.RawInterval
-			req.AggNum = req.OutInterval / req.RawInterval
+		if req.ArchInterval == interval {
+			// the easy case. we can satisfy this req with what we already found
+			// just have to set a few more options
+			req.OutInterval = req.ArchInterval
+			req.AggNum = 1
+
+		} else {
+			// the harder case. due to other reqs with different retention settings
+			// we have to deliver an interval higher than what we originally came up with
+
+			// let's see first if we can deliver it via lower-res rollup archives, if we have any
+			retentions := mdata.GetRetentions(req.SchemaI)
+			for i, ret := range retentions[req.Archive+1:] {
```

Comment:
if `!ret.Ready` then the retention will not be useful here, so we might as well do `if !ret.Ready {continue}` and skip the potential unnecessary `uint32(ret.SecondsPerPoint())`.